extensions/ql-vscode/src/remote-queries/analyses-results-manager.ts

```diff
@@ -0,0 +1,90 @@
+import { ExtensionContext } from 'vscode';
+import { Credentials } from '../authentication';
+import { Logger } from '../logging';
+import { downloadArtifactFromLink } from './gh-actions-api-client';
+import { DownloadLink } from './download-link';
+import * as path from 'path';
+import * as fs from 'fs-extra';
+import { AnalysisSummary } from './shared/remote-query-result';
+import * as sarif from 'sarif';
+import { AnalysisResults, QueryResult } from './shared/analysis-result';
+
+export class AnalysesResultsManager {
+  // Store for the results of various analyses for a single remote query.
+  private readonly analysesResults: { [key: string]: QueryResult[] };
+
+  constructor(
+    private readonly ctx: ExtensionContext,
+    private readonly logger: Logger,
+  ) {
+    this.analysesResults = {};
+  }
+
+  public async downloadAnalysisResults(
+    analysisSummary: AnalysisSummary,
+  ): Promise<void> {
+    const credentials = await Credentials.initialize(this.ctx);
+
+    void this.logger.log(`Downloading and processing results for ${analysisSummary.nwo}`);
+
+    const queryResults = await this.downloadSingleAnalysisResults(
+      analysisSummary.downloadLink,
+      credentials);
+
+    this.analysesResults[analysisSummary.nwo] = queryResults;
+  }
+
+  public async downloadAllResults(
+    analysisSummaries: AnalysisSummary[],
+  ): Promise<void> {
+    const credentials = await Credentials.initialize(this.ctx);
+
+    void this.logger.log('Downloading and processing all results');
+
+    for (const analysis of analysisSummaries) {
+      const queryResults = await this.downloadSingleAnalysisResults(analysis.downloadLink, credentials);
+      this.analysesResults[analysis.nwo] = queryResults;
```

Comment:
This downloads each analysis serially, which could take some time if there are lots of them. Should we be parallelizing the requests? We probably don't want to parallelize all requests, but possibly parallelize slices of requests.

This is not something to work on now, only a consideration for the future.

Reply:
Ah sorry I was planning to leave a comment here but I forgot! I was thinking that some parallelisation would be a good idea too but was unsure about how many concurrent downloads we should be doing. I think the number needs to take into account that people may be running a query against a lot of repos, some with potentially big result files and some with potentially small ones. I was thinking we could start with 3 and see how it goes. What do you think?

Agree we could leave it as is for now - I'll be working on auto-downloads etc. too so might be worth addressing it then.

Reply:
That all sounds reasonable. 3 is a good place to start.